nifs/atomvm_adc.c

```diff
@@ -15,93 +15,71 @@
-#include <esp_log.h>
-#include <driver/adc.h>
-#ifdef CONFIG_AVM_ADC2_ENABLE
-#  if defined __has_include
-#    if __has_include (<driver/adc2_wifi_private.h>)
-#      include <driver/adc2_wifi_private.h>
-#    elif  __has_include (<driver/adc2_wifi_internal.h>)
-#      include <driver/adc2_wifi_internal.h>
-#    endif
-#  endif
-#endif
 #include <context.h>
 #include <defaultatoms.h>
-#include <esp32_sys.h>
 #include <interop.h>
 #include <nifs.h>
 #include <term.h>
 
-#include "atomvm_adc.h"
+#define ENABLE_TRACE
+#include <trace.h>
 
-//#define ENABLE_TRACE
-#include "trace.h"
+#include <driver/adc.h>
+#include <esp32_sys.h>
+#include <esp_adc_cal.h>
+#include <esp_log.h>
+#include <sdkconfig.h>
+#include <stdlib.h>
+
+#include "atomvm_adc.h"
 
 #define TAG "atomvm_adc"
 #define DEFAULT_SAMPLES 64
-#define DEFAULT_VREF    1100
-
-// References
-// https://docs.espressif.com/projects/esp-idf/en/v3.3.4/api-reference/peripherals/adc.html
-// https://docs.espressif.com/projects/esp-idf/en/v4.2.3/api-reference/peripherals/adc.html << v4.2 switch from adc2_wifi_internal.h to adc2_wifi_private.h
-// https://docs.espressif.com/projects/esp-idf/en/v4.4.2/api-reference/peripherals/adc.html
-//
-
-#if CONFIG_IDF_TARGET_ESP32
-static const char *const bit_9_atom           = "\x5"  "bit_9";
-static const char *const bit_10_atom          = "\x6"  "bit_10";
-static const char *const bit_11_atom          = "\x6"  "bit_11";
-static const char *const bit_12_atom          = "\x6"  "bit_12";
+#define DEFAULT_VREF 1100
+
+static const AtomStringIntPair bit_width_table[] = {
+    { ATOM_STR("\x7", "bit_max"), (ADC_WIDTH_BIT_DEFAULT) },
+#if SOC_ADC_MAX_BITWIDTH == 13
+    { ATOM_STR("\x6", "bit_13"), ADC_WIDTH_BIT_13 },
+#elif SOC_ADC_MAX_BITWIDTH == 12
+    { ATOM_STR("\x6", "bit_12"), ADC_WIDTH_BIT_12 },
+#elif CONFIG_IDF_TARGET_ESP32
+    { ATOM_STR("\x6", "bit_11"), ADC_WIDTH_BIT_11 },
+    { ATOM_STR("\x6", "bit_10"), ADC_WIDTH_BIT_10 },
+    { ATOM_STR("\x5", "bit_9"), ADC_WIDTH_BIT_9 },
 #endif
+    SELECT_INT_DEFAULT(ADC_WIDTH_MAX)
+};
 
-static const char *const db_0_atom            = "\x4"  "db_0";
-static const char *const db_2_5_atom          = "\x6"  "db_2_5";
-static const char *const db_6_atom            = "\x4"  "db_6";
-static const char *const db_11_atom           = "\x5"  "db_11";
+static const AtomStringIntPair attenuation_table[] = {
+    { ATOM_STR("\x4", "db_0"), ADC_ATTEN_DB_0 },
+    { ATOM_STR("\x6", "db_2_5"), ADC_ATTEN_DB_2_5 },
+    { ATOM_STR("\x4", "db_6"), ADC_ATTEN_DB_6 },
+    { ATOM_STR("\x6", "db_11"), ADC_ATTEN_DB_11 },
+    SELECT_INT_DEFAULT(ADC_ATTEN_MAX)
+};
 
-static const char *const samples_atom         = "\x7"  "samples";
-static const char *const raw_atom             = "\x3"  "raw";
-static const char *const voltage_atom         = "\x7"  "voltage";
+static const char *const samples_atom = "\x7" "samples";
+static const char *const raw_atom     = "\x3" "raw";
+static const char *const voltage_atom = "\x7" "voltage";
```

Comment:
Can we use the ATOM_STR macro here, while we are at it?

Reply:
Yes! My mistake... I had added them, and then ran clang-format in place over the file, which did some nasty things to those lines. So I used git to revert then, and it looks like I forgot to commit those changes before running the format...

Reply:
Actually, I was just asking if you could use, for example,
```
ATOM_STR("\x7", "voltage")
```

which would address the reformatting stuff that happens to me, too.